photon-lib/src/main/scala/com/linkedin/photon/ml/data/MinHeapWithFixedCapacity.scala

```diff
@@ -16,42 +16,71 @@ package com.linkedin.photon.ml.data
 
 import java.{util => Jutil}
 
-import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
 import scala.collection.mutable
 import scala.reflect.ClassTag
 
 /**
- * Implementation of a min heap data structure that defers heapification until the given buffer capacity has been
+ * Implementation of a min heap data structure that defers 'heapification' until the given buffer capacity has been
  * reached.
  *
  * @param capacity Buffer capacity
  */
 protected[ml] class MinHeapWithFixedCapacity[T <: Comparable[T] : ClassTag](capacity: Int) extends Serializable {
 
+  // Buffer to use until capacity is reached
+  // TODO: Better size hint for buffer?
   private val arrayBuffer = new mutable.ArrayBuffer[T](2)
-  var cumCount = 0
+  // Min heap to use once buffer size exceeded
   private var minHeap: Jutil.PriorityQueue[T] = _
+  // The number of items that have been attempted to be added to the heap
+  private var cumCount = 0
 
   /**
-   * Add value to the heap.
+   * Get the total number of items which have been attempted to be added to the min heap (note that this may be more
+   * than the number of items currently in the min heap, since it has fixed capacity).
+   *
+   * @return The number of items added to the min heap
+   */
+  def getCount: Int = cumCount
+
+  /**
+   * Return the heap data.
+   *
+   * @return The heap data
+   */
+  def getData: Iterable[T] = {
+
+    if (cumCount < capacity) {
+      arrayBuffer
+    } else {
+      minHeap.asScala
+    }
+  }
+
+  /**
+   * Add a value to the heap.
    *
    * @param value The value to add
    * @return The updated heap
    */
   def +=(value: T): this.type = {
+
+    // If capacity not reached, add to buffer
     if (cumCount < capacity) {
-      arrayBuffer.add(value)
-    } else {
-      if (minHeap == null) activateMinHeap()
-      if (minHeap.peek().compareTo(value) <= 0) {
-        minHeap.poll()
-        minHeap.offer(value)
-      }
+      arrayBuffer += value
+
+    // Otherwise: if new item greater than smallest item in heap, replace the heap item with the new item
+    } else if (minHeap.peek().compareTo(value) <= 0) {
```

Comment:
What if `minHeap` is null?

Reply:
This can only occur if the heap has not been activated when the first item is added and the capacity is 0. Adding restriction that capacity must be greater than 0.